### DISCUSSION TITLE
Is there a way to query the server directly without the front end?

### DISCUSSION THREAD

**Discussion by grantbrewster** [rom1504/clip-retrieval#86] [Q&A] [answered]
Hey! I was wondering if there is some way to get images from querying the server with a request directly from code instead of through the front end. Love your work, hoping to incorporate it into a class project.

**Comment by rom1504** [comment, marked as answer]
Hi, yes you can either load the knn index yourself (see the filter notebook) or you can all our backend (the API is described in the readme) for example using request lib in python

**Reply by grantbrewster** [reply]
Got it, thank you!

**Comment by xilongzhou** [comment]
Hi Could you please give me some suggestion how to use the simplified_filter ? I check the file but confused about a couple of things:
1. what is `"/home/rom1504/indices_paths.json"`? my understanding is that it should be index results containing image.index, text.index and metadata folder. but if I want to filter over Laion5B like frontend, where should we get this? I find this [link](https://huggingface.co/datasets/laion/laion5B-index/tree/main) but it does not work for me.

2. I got some error `load_clip_indices() takes 2 positional arguments but 5 were given`, seems like the current version does not match this filter file?

Thank you very much!

**Reply by rom1504** [reply]
Hi, it is explained in the readme how to write the configuration file

I do not advise using clip filter for large indices

Instead you can load the index and metadata using clip back API and do the querying you want yourself directly

**Reply by xilongzhou** [reply]
Let me try. Thanks for your help!

**Comment by zeke** [comment]
@afiaka87 just landed PR https://github.com/rom1504/clip-retrieval/pull/157 which I think is related to this question.

**Reply by xilongzhou** [reply]
Great! I tried clip client and it works pretty well ! However, I came across an issue related to the number of images: 

When I initialize `ClipClient` with different `num_images`, I realize that the number of images API actually return does not match this number (for example, if I set number as 20k, it returns 3k images) and when I set this number too large (>30k), there will be some error caused by time-out. So there is certain upper limit for this number, right?

I appreciate your help!

**Reply by rom1504** [reply]
Post filtering decrease the number of results that is actually retrieved : (approximate) deduplication per image and safety filtering. They also both slow things down
You can disable them both if you want to fetch more results

However, if you want to fetch a large subset i would currently recommend to do it locally.

**Reply by rom1504** [reply]
https://github.com/rom1504/clip-retrieval/issues/160 will be what needs to be done to be able to increase num images much more

**Reply by xilongzhou** [reply]
Got it. Thanks!

**Comment by pengfeiwu1999** [comment]
I'm sorry because I'm a fresh guy to use the CLIP model, so I just have a question about your backend part: how can I change the parameter "url" in "client = ClipClient(url="https://knn5.laion.ai/knn-service", indice_name="laion5B") " to change the backbone of your backend backbone（such as "ViT-B/32" and others）

**Reply by rom1504** [reply]
For that you need to run the backend yourself.
Clip embeddings indexed in the index are for a fixed model (that's ViT-l/14 for laion5B)

**Reply by pengfeiwu1999** [reply]
> 

“run the backend yourself” means that I should construct a backbone to do the inference and retrieval by myself locally rather than use your clipclient interface directly?

**Reply by rom1504** [reply]
Yes, you can check the readme to find out how

**Reply by pengfeiwu1999** [reply]
> 

thanks a lot for your contribution
wonderful repo!!!
